duckdb-vortex/src/vortex_extension.cpp

```diff
@@ -49,51 +50,82 @@ struct VortexBindData : public TableFunctionData {
 /// Local state for the Vortex table function that tracks the progress of a scan
 /// operation. In DuckDB's execution model, a query reading from a file can be
 /// parallelized by dividing it into ranges, each handled by a different scan.
-struct VortexScanState : public LocalTableFunctionState {
+struct VortexScanLocalState : public LocalTableFunctionState {
 	idx_t current_row = 0;
 	bool finished = false;
 	mutable Array *array;
+};
+
+struct VortexScanGlobalState : public GlobalTableFunctionState {
+	ArrayStream *array_stream;
+	std::mutex stream_lock;
+	bool finished;
 
 	optional_ptr<TableFilterSet> filter;
 	// The column idx that must be returned by the scan.
 	vector<idx_t> column_ids;
-};
 
-struct VortexScanSharedState : public GlobalTableFunctionState {
+	vector<idx_t> projection_ids;
+
 	idx_t MaxThreads() const override {
-		return 1;
+		return 99999;
 	}
 };
 
 static void VortexScanFunction(ClientContext &context, TableFunctionInput &data, DataChunk &output) {
-	auto &bind_data = data.bind_data->Cast<VortexBindData>(); // NOLINT
-	auto &state = data.local_state->Cast<VortexScanState>();  // NOLINT
-
-	if (bind_data.array_stream == nullptr) {
-		const char **const column_names = new const char *[state.column_ids.size()];
-		auto idx = 0;
-		for (auto col_id : state.column_ids) {
-			column_names[idx++] = bind_data.column_names[col_id].c_str();
-		}
-		auto options = FileScanOptions {column_names, static_cast<int>(state.column_ids.size())};
-
-		bind_data.array_stream = File_scan(bind_data.file, &options);
-
-		delete[] column_names;
-	}
+	auto &bind_data = data.bind_data->Cast<VortexBindData>();         // NOLINT
+	auto &state = data.local_state->Cast<VortexScanLocalState>();     // NOLINT
+	auto &g_state = data.global_state->Cast<VortexScanGlobalState>(); // NOLINT
 
 	if (state.finished) {
 		return;
 	}
 
 	if (state.array == nullptr) {
-		auto next = FFIArrayStream_next(bind_data.array_stream);
+		std::lock_guard l(g_state.stream_lock);
+
+		if (g_state.finished) {
```

Comment:
Should `local_state.finished` be set true in this branch?

Reply:
I can, but there is no need since I will never be called again

Reply:
Can we check for if global state or local state finished at the top of the function, in one go?

Or asked differently, should this run even if global state finished == true?
```

	state.current_row =
	    FFIArray_to_duckdb_chunk(state.array, state.current_row, reinterpret_cast<duckdb_data_chunk>(&output));

	if (state.current_row == 0) {
		FFIArray_free(state.array);
		state.array = nullptr;
	}
```